mgm/src/lib.rs

```diff
@@ -172,190 +116,100 @@ where
-        }
-
-        let mut enc_ctr = nonce.clone();
-        enc_ctr[0] &= 0b0111_1111;
-        self.cipher.encrypt_block(&mut enc_ctr);
-        let mut enc_ctr = C::BlockSize::block2ctr(&enc_ctr);
-
-        // process plaintext
-        if pb > 1 {
-            let mut iter = buffer.chunks_exact_mut(pb * bs);
-            for chunk in &mut iter {
-                self.apply_par_ks_blocks(&mut enc_ctr, chunk);
-                self.update_par_tag(&mut tag, &mut tag_ctr, chunk);
-            }
-            buffer = iter.into_remainder();
-        }
-
-        let mut iter = buffer.chunks_exact_mut(bs);
-        for block in (&mut iter).map(Block::<C>::from_mut_slice) {
-            self.apply_ks_block(&mut enc_ctr, block);
-            self.update_tag(&mut tag, &mut tag_ctr, block);
-        }
-        let rem = iter.into_remainder();
-        if !rem.is_empty() {
-            self.apply_ks_block(&mut enc_ctr, rem);
-
-            let mut block = Block::<C>::default();
-            let n = rem.len();
-            block[..n].copy_from_slice(rem);
-            self.update_tag(&mut tag, &mut tag_ctr, &block);
-        }
-
-        let block = C::BlockSize::lengths2block(adata_len, msg_len)?;
-        self.update_tag(&mut tag, &mut tag_ctr, &block);
-
-        let mut tag = tag.into_bytes();
-        self.cipher.encrypt_block(&mut tag);
-
-        Ok(tag)
+        mgm_encrypt((&self.cipher, nonce, adata, buffer))
     }
 
     fn decrypt_in_place_detached(
         &self,
         nonce: &Nonce<Self::NonceSize>,
-        mut adata: &[u8],
-        mut buffer: &mut [u8],
+        adata: &[u8],
+        buffer: &mut [u8],
         expected_tag: &Tag<Self::TagSize>,
     ) -> Result<(), Error> {
         // first nonce bit must be equal to zero
         if nonce[0] >> 7 != 0 {
             return Err(Error);
         }
+        mgm_decrypt((&self.cipher, nonce, adata, buffer, expected_tag))
+    }
+}
 
-        let adata_len = adata.len();
-        let msg_len = buffer.len();
-
-        let mut tag_ctr = nonce.clone();
-        tag_ctr[0] |= 0b1000_0000;
-        self.cipher.encrypt_block(&mut tag_ctr);
-
-        let mut tag_ctr = C::BlockSize::block2ctr(&tag_ctr);
-        let mut tag = <C::BlockSize as Sealed>::Element::new();
-
-        let pb = C::ParBlocks::USIZE;
-        let bs = C::BlockSize::USIZE;
-
-        // calculate tag
-        // process adata
-        if pb > 1 {
-            let mut iter = adata.chunks_exact(pb * bs);
-            for chunk in &mut iter {
-                self.update_par_tag(&mut tag, &mut tag_ctr, chunk);
+cfg_if! {
+    if #[cfg(all(
+        any(target_arch = "x86_64", target_arch = "x86"),
+        not(feature = "force-soft")
+    ))] {
+        #[target_feature(enable = "pclmulqdq")]
+        #[target_feature(enable = "ssse3")]
+        #[target_feature(enable = "sse2")]
+        unsafe fn wrapper<R: Sized>(f: impl FnOnce() -> R) -> R {
+            f()
+        }
```

Comment:
I was not sure if this trick will work, but by inspecting generated assembly it looks like it does. Although I am not sure how "deep" can it work, e.g. I don't think it will influence code generation of items coming from different crates, unless they get inlined.